pkg/operation/botanist/secrets.go

```diff
@@ -1001,6 +974,90 @@ func (b *Botanist) generateShootSecrets(ctx context.Context, existingSecretsMap
 	return nil
 }
 
+// DeployEtcdSecret creates etcd main and events tls secrets. It is different from DeploySecrets
+// as we shall get the names of the etcd instances only after it is deployed.
+func (b *Botanist) DeployEtcdSecret(ctx context.Context) error {
+	existingSecretsMap, err := b.fetchExistingSecrets(ctx)
+	if err != nil {
+		return err
+	}
+
+	certificateAuthorities, err := b.generateCertificateAuthorities(existingSecretsMap)
+	if err != nil {
+		return err
+	}
+
+	etcds, err := b.fetchEtcds(ctx)
+	if err != nil {
+		return err
+	}
+
+	etcdSecretList := b.generateEtcdSecrets(certificateAuthorities, etcds)
+	if err := b.generateShootSecrets(ctx, existingSecretsMap, etcdSecretList); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (b *Botanist) fetchEtcds(ctx context.Context) (map[string]string, error) {
+	etcdList := &druidv1alpha1.EtcdList{}
+	if err := b.K8sSeedClient.Client().List(ctx, etcdList, client.InNamespace(b.Shoot.SeedNamespace)); err != nil {
+		return nil, err
+	}
+	etcds := map[string]string{}
+	for _, etcd := range etcdList.Items {
+		if etcd.Name == v1alpha1constants.StatefulSetNameETCDMain {
+			if etcd.Status.Etcd.Name == "" {
+				return nil, fmt.Errorf("etcd statefulset has not yet been populated in etcd %s", v1alpha1constants.StatefulSetNameETCDMain)
+			}
+			etcds[v1alpha1constants.StatefulSetNameETCDMain] = etcd.Status.Etcd.Name
+
+		} else if etcd.Name == v1alpha1constants.StatefulSetNameETCDEvents {
+			if etcd.Status.Etcd.Name == "" {
+				return nil, fmt.Errorf("etcd statefulset has not yet been populated in etcd %s", v1alpha1constants.StatefulSetNameETCDMain)
+			}
+			etcds[v1alpha1constants.StatefulSetNameETCDEvents] = etcd.Status.Etcd.Name
+		}
+	}
+	return etcds, nil
+}
+
+func (b *Botanist) generateEtcdSecrets(certificateAuthorities map[string]*secrets.Certificate, etcds map[string]string) []secrets.ConfigInterface {
+	etcdCertDNSNames := dnsNamesForEtcd(map[string]string{
+		v1alpha1constants.StatefulSetNameETCDMain:   etcds[v1alpha1constants.StatefulSetNameETCDMain],
```

Comment:
I don't understand this code here. Why do we do all of this and don't just keep how it was before? AFAIK, we still need the exact same secrets for etcd, so why touching this logic now? What is the motivation of listing the `Etcd` resources? This line, especially, will probably panic with a nil-pointer exception for existing clusters, or?

Reply:
Etcd statefulset pod had the form etcd-main-0 as the alternate name in the certificate. With druid, the statefulset will add a nonce to the statefulset name. So the pod name will be like etcd-main-nonce-0. We need to pull the name of the etcd statefulset from the status of the etcd resource.

Reply:
It will not panic for existing clusters as `etcd-druid` adopts existing etcd statefulsets and return `etcd-main-0` for etcd statefulset pod.